File: src/main/java/org/mastodon/mamut/util/GroupLocksHandling.java
package org.mastodon.mamut.util;

import org.mastodon.app.ui.GroupLocksPanel;
import org.mastodon.grouping.GroupHandle;
import org.mastodon.mamut.MamutAppModel;
import org.mastodon.mamut.SciviewBridge;
import org.mastodon.mamut.WindowManager;
import org.mastodon.mamut.model.Link;
import org.mastodon.mamut.model.Spot;
import org.mastodon.model.NavigationListener;
import org.mastodon.model.TimepointListener;
import org.mastodon.pool.PoolCollectionWrapper;
import org.scijava.AbstractContextual;
import org.scijava.event.EventHandler;
import org.scijava.event.EventSubscriber;
import org.scijava.event.SciJavaEvent;
import sc.iview.event.NodeActivatedEvent;
import java.util.List;
import java.util.Optional;

public class GroupLocksHandling {
	private final SciviewBridge bridge;    //controls sciview via this bridge obj
	private final MamutAppModel appModel;  //controls Mastodon
	private final PoolCollectionWrapper<Spot> vertices; //shortcut to inside of Mastodon
	private final NavigationRequestsHandler navigationRequestsHandler = new NavigationRequestsHandler();
	private final SciviewEventListener sciviewFocusHandler = new SciviewEventListener();

	public GroupLocksHandling(final SciviewBridge bridge, final WindowManager mastodon) {
		this.bridge = bridge;
		this.appModel = mastodon.getAppModel();
		this.vertices = appModel.getModel().getGraph().vertices();
		this.isActive = false; //'cause it's not having any group handle yet
	}


	private GroupHandle myGroupHandle;
	private boolean isActive;

	public GroupLocksPanel createAndActivate() {
		if (isActive) return null;
		isActive = true;

		bridge.getEventService().subscribe(sciviewFocusHandler);

		myGroupHandle = appModel.getGroupManager().createGroupHandle();
		myGroupHandle.getModel(appModel.NAVIGATION).listeners().add(navigationRequestsHandler);
		myGroupHandle.getModel(appModel.TIMEPOINT).listeners().add(navigationRequestsHandler);
		return new GroupLocksPanel(myGroupHandle);
	}

	public void deactivate() {
		if (!isActive) return;
		isActive = false;

		myGroupHandle.getModel(appModel.NAVIGATION).listeners().remove(navigationRequestsHandler);
		myGroupHandle.getModel(appModel.TIMEPOINT).listeners().remove(navigationRequestsHandler);
		appModel.getGroupManager().removeGroupHandle(myGroupHandle);

		//can fail, so we better do it as the last action here
		List<EventSubscriber<SciJavaEvent>> subs = bridge.getEventService().getSubscribers(SciJavaEvent.class);
		if (subs != null) subs.remove(sciviewFocusHandler);
	}


	class NavigationRequestsHandler implements NavigationListener<Spot,Link>, TimepointListener {
		@Override
		public void navigateToVertex(Spot vertex) {
			if (isActive) focusSciviewToNode(vertex.getLabel());
		}
		@Override
		public void navigateToEdge(Link edge) {
			if (isActive) focusSciviewToNode(edge.getSource().getLabel());
		}

		@Override
		public void timepointChanged() {
			bridge.showTimepoint( myGroupHandle.getModel(appModel.TIMEPOINT).getTimepoint() );
		}
	}

	class SciviewEventListener extends AbstractContextual {
		@EventHandler
		public void onEvent(NodeActivatedEvent event) {
			if (event.getNode() == null) return;
			if (isActive) focusMastodonToSpot(event.getNode().getName());
		}
	}

	public void focusMastodonToSpot(final String name) {
		Optional<Spot> res = vertices.stream().filter( s -> name.matches(s.getLabel()) ).findFirst();
		if (res.isPresent()) {
			//TODO: this is triggering also our handler (see above navigateToVertex()) !
			myGroupHandle.getModel(appModel.NAVIGATION).notifyNavigateToVertex(res.get());
		}
	}

	public void focusSciviewToNode(final String name) {
		bridge.focusSpot(name);
	}
}

File: src/main/kotlin/util/GroupLocksHandling.kt
package util

import org.mastodon.app.ui.GroupLocksPanel
import org.mastodon.grouping.GroupHandle
import org.mastodon.mamut.ProjectModel
import org.mastodon.mamut.SciviewBridge
import org.mastodon.mamut.model.Link
import org.mastodon.mamut.model.Spot
import org.mastodon.model.NavigationListener
import org.mastodon.model.TimepointListener
import org.mastodon.pool.PoolCollectionWrapper
import org.scijava.AbstractContextual
import org.scijava.event.EventHandler
import org.scijava.event.SciJavaEvent
import sc.iview.event.NodeActivatedEvent

class GroupLocksHandling(//controls sciview via this bridge obj
    private val bridge: SciviewBridge, mastodon: ProjectModel
) {
    //controls Mastodon
    private val projectModel: ProjectModel = mastodon
    //shortcut to inside of Mastodon
    private val vertices: PoolCollectionWrapper<Spot> = projectModel.model.graph.vertices()
    private val navigationRequestsHandler: NavigationRequestsHandler = NavigationRequestsHandler()
    private val sciviewFocusHandler: SciviewEventListener = SciviewEventListener()
    private lateinit var myGroupHandle: GroupHandle
    private var isActive = false

    fun createAndActivate(): GroupLocksPanel? {
        if (isActive) return null
        isActive = true
        bridge.eventService?.subscribe(sciviewFocusHandler)
        myGroupHandle = projectModel.groupManager.createGroupHandle()
        myGroupHandle.getModel(projectModel.NAVIGATION).listeners().add(navigationRequestsHandler)
        myGroupHandle.getModel(projectModel.TIMEPOINT).listeners().add(navigationRequestsHandler)
        return GroupLocksPanel(myGroupHandle)
    }

    fun deactivate() {
        if (!isActive) return
        isActive = false
        myGroupHandle.getModel(projectModel.NAVIGATION).listeners().remove(navigationRequestsHandler)
        myGroupHandle.getModel(projectModel.TIMEPOINT).listeners().remove(navigationRequestsHandler)
        projectModel.groupManager.removeGroupHandle(myGroupHandle)

        //can fail, so we better do it as the last action here
        val subs = bridge.eventService?.getSubscribers(
            SciJavaEvent::class.java
        )
        subs?.remove<Any?>(sciviewFocusHandler)
    }

    internal inner class NavigationRequestsHandler : NavigationListener<Spot, Link>, TimepointListener {
        override fun navigateToVertex(vertex: Spot) {
            if (isActive) focusSciviewToNode(vertex.label)
        }

        override fun navigateToEdge(edge: Link) {
            if (isActive) focusSciviewToNode(edge.source.label)
        }

        override fun timepointChanged() {
            println("timepointChanged to ${myGroupHandle.getModel(projectModel.TIMEPOINT).timepoint}")
            bridge.showTimepoint(myGroupHandle.getModel(projectModel.TIMEPOINT).timepoint)
        }
    }

    internal inner class SciviewEventListener : AbstractContextual() {
        @EventHandler
        fun onEvent(event: NodeActivatedEvent) {
            if (event.node == null) return
            if (isActive) focusMastodonToSpot(event.node.name)
        }
    }

    fun focusMastodonToSpot(name: String) {
        val res = vertices.stream().filter { s: Spot -> name.matches(s.label.toRegex()) }.findFirst()
        if (res.isPresent) {
            //TODO: this is triggering also our handler (see above navigateToVertex()) !
            myGroupHandle.getModel(projectModel.NAVIGATION).notifyNavigateToVertex(res.get())
        }
    }

    fun focusSciviewToNode(name: String) {
        bridge.focusSpot(name)
    }
}
